agent/consul/gateways/controller_gateways.go

```diff
@@ -2,33 +2,202 @@ package gateways
 
 import (
 	"context"
-
-	"github.com/hashicorp/go-hclog"
-
 	"github.com/hashicorp/consul/agent/consul/controller"
-	"github.com/hashicorp/consul/agent/consul/fsm"
+	"github.com/hashicorp/consul/agent/consul/gateways/datastore"
 	"github.com/hashicorp/consul/agent/consul/state"
 	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/go-hclog"
+	"github.com/pkg/errors"
 )
 
 type apiGatewayReconciler struct {
-	fsm    *fsm.FSM
 	logger hclog.Logger
+	store  datastore.DataStore
 }
 
-func (r apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
-	return nil
-}
-
-func NewAPIGatewayController(fsm *fsm.FSM, publisher state.EventPublisher, logger hclog.Logger) controller.Controller {
+// NewAPIGatewayController returns a new APIGateway controller
+func NewAPIGatewayController(store datastore.DataStore, publisher state.EventPublisher, logger hclog.Logger) controller.Controller {
 	reconciler := apiGatewayReconciler{
-		fsm:    fsm,
 		logger: logger,
+		store:  store,
 	}
-	return controller.New(publisher, reconciler).Subscribe(
+	return controller.New(publisher, &reconciler).Subscribe(
 		&stream.SubscribeRequest{
 			Topic:   state.EventTopicAPIGateway,
 			Subject: stream.SubjectWildcard,
 		},
 	)
 }
+
+// Reconcile takes in a controller request and ensures this api gateways corresponding BoundAPIGateway exists and is
+// up to date
+func (r *apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
+	metaGateway, err := r.initGatewayMeta(req)
+	if err != nil {
+		return err
+	} else if metaGateway == nil && err == nil {
```

Comment:
I don't think we need the `&& err == nil` here because it can't be non-nil here from the previous condition, we can also probably just make this `if metaGateway == nil` and not need the `else if` because we're exiting early if `err != nil`